### PR TITLE
A followup to #502 and merge some internal changes.

### PR DESCRIPTION
**Pull request by mingxingtan**
Important changes:
1. Changed config to allow hierarchical dict or configs.

Minor changes:
2. Deleted some unless configs.
3. Added tests for focal loss.
4. Fixed some format and coding style issues.

### CODE REVIEW

File: efficientdet/det_model_fn.py [hunk 1]
```diff
@@ -142,7 +142,24 @@ def learning_rate_schedule(params, global_step):
   raise ValueError('unknown lr_decay_method: {}'.format(lr_decay_method))
 
 
-def focal_loss(y_pred, y_true, alpha, gamma, label_smoothing, normalizer):
+def legacy_focal_loss(logits, targets, alpha, gamma, normalizer, _):
```
Comment (mingxingtan) [inline efficientdet/det_model_fn.py:145]: I kept the original focal loss here for now, such that we can switch the loss function if we see any issues in near future. 

We can remove this legacy_focal_loss in future after we fully verified the new loss works well as expected.

File: efficientdet/det_model_fn.py [hunk 2]
```diff
@@ -156,24 +173,26 @@ def focal_loss(y_pred, y_true, alpha, gamma, label_smoothing, normalizer):
...
 
-    # Get the cross_entropy for each entry
-    ce = tf.keras.losses.categorical_crossentropy(
-        y_true, y_pred, from_logits=True, label_smoothing=label_smoothing)
+    # apply label smoothing.
+    y_true = y_true * (1.0 - label_smoothing) + 0.5 * label_smoothing
 
-    pred_prob = tf.sigmoid(y_pred)
+    # get cross_entropy for each entry.
+    ce = tf.nn.sigmoid_cross_entropy_with_logits(labels=y_true, logits=y_pred)
```
Comment (mingxingtan) [inline efficientdet/det_model_fn.py:190]: tf.keras.losses.categorical_crossentropy will change the output shape, so I have to use tf.nn.sigmoid_cross_entropy_with_logits  here.

File: efficientdet/det_model_fn_test.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+# ==============================================================================
+"""Tests for det_model_fn."""
+import tensorflow as tf
+
+import det_model_fn
+
+
+class DetModelFnTest(tf.test.TestCase):
+
+  def test_focal_loss(self):
```
Comment (mingxingtan) [inline efficientdet/det_model_fn_test.py:24]: @Ely-S I added a simple test to ensure the new focal_loss function has the same output if label_smoothing=0.  Could you also add a test to ensure the output is expected when label_smoothing != 0?

File: efficientdet/efficientdet_arch.py [hunk 4]
```diff
@@ -154,46 +131,28 @@ def _maybe_apply_1x1(feat):
...
-        else:
-          feat = nearest_upsampling(
-              feat,
-              height_scale=height_scale,
-              width_scale=width_scale,
-              data_format=data_format)
+        if data_format == 'channels_first':
+          feat = tf.transpose(feat, [0, 2, 3, 1])
+        feat = tf.cast(
+            tf.image.resize_nearest_neighbor(
```
Comment (mingxingtan) [inline efficientdet/efficientdet_arch.py:151]: I deleted the customized resize_nearest_neighbor (it was for some legacy issues, not used in the current code).

File: efficientdet/hparams_config.py [hunk 5]
```diff
@@ -132,16 +123,34 @@ def save_to_yaml(self, yaml_file_path):
       yaml.dump(self.as_dict(), f, default_flow_style=False)
 
   def parse_from_str(self, config_str: Text) -> Dict[Any, Any]:
-    """parse from a string in format 'x=a,y=2' and return the dict."""
+    """Parse a string like 'x.y=1,x.z=2' to nested dict {x: {y: 1, z: 2}}."""
```
Comment (mingxingtan) [inline efficientdet/hparams_config.py:126]: This file contains some changes to allow nested configs.

File: efficientdet/keras/efficientdet_arch_keras.py [hunk 6]
```diff
@@ -351,8 +325,6 @@ def get_config(self):
         'target_width': self.target_width,
```
Comment (fsx950223) [inline efficientdet/keras/efficientdet_arch_keras.py:325]: We could remove ```get_config``` if we don't use keras model.

File: efficientdet/inference.py [hunk 7]
```diff
@@ -255,14 +255,14 @@ def det_post_process_combined(params, cls_outputs, box_outputs, scales,
   image_ids = tf.cast(
       tf.tile(
           tf.expand_dims(tf.range(batch_size), axis=1), [1, max_boxes_to_draw]),
-      dtype=tf.float32)
+      dtype=scales.dtype)
```
Comment (fsx950223) [inline efficientdet/inference.py:258]: Why?

File: efficientdet/keras/efficientdet_arch_keras.py (hunk 6 above)
Comment (mingxingtan) [reply efficientdet/keras/efficientdet_arch_keras.py:325]: I will keep this for now since this PR is already large.

File: efficientdet/inference.py (hunk 7 above)
Comment (mingxingtan) [reply efficientdet/inference.py]: Oops, it was an old change. Reverted now.

File: efficientdet/det_model_fn.py (hunk 2 above)
Comment (Ely-S) [reply efficientdet/det_model_fn.py:190]: :+1:

**Comment by Ely-S** [comment]
Hi @mingxingtan, I've been travelling all week and didn't get to finish my last PR or this review one, but it looks great to me!

File: efficientdet/anchors.py [hunk 8]
```diff
@@ -115,15 +113,22 @@ def decode_box_outputs_tf(rel_codes, anchors):
...
      / square of diagonal of smallest enclosing bounding box
 
   Reference: https://arxiv.org/pdf/1911.08287.pdf
+
+  Args:
+    dets: detection with shape (num, 5) and format [x1, y1, x2, y2, score].
+    iou_thresh: IOU threshold,
+
+  Returns:
+    numpy.array: Retained boxes.
```
Comment (Ely-S) [inline efficientdet/anchors.py:129]: :+1:  good docstring

File: efficientdet/det_model_fn.py (hunk 1 above)
Comment (Ely-S) [reply efficientdet/det_model_fn.py:145]: :+1:
